string/Cursor.ts

```diff
@@ -0,0 +1,61 @@
+/** Describes a string and an index within it. */
+export class Cursor {
+  #input: { value: string };
+
+  constructor(
+    input: string,
+    public index: number = 0,
+  ) {
+    this.#input = { value: input };
+  }
+
+  /** Create a new Cursor with the same input and index. */
+  static copy(input: Cursor): Cursor {
+    const cursor = new Cursor("", input.index);
+    cursor.#input = input.#input;
+
+    return cursor;
+  }
+
+  /** An immutable reference to the input string. */
+  get input(): string {
+    return this.#input.value;
+  }
+
+  /** The character at the current index, if extant. */
+  get char(): string | undefined {
+    return this.#input.value.at(this.index);
+  }
+
+  /** The substring from the start of the string to the current index. */
+  get antecedent(): string {
+    return this.index === 0 ? "" : this.#input.value.slice(0, this.index);
+  }
+
+  /** The substring from the current index to the end of the input. */
+  get remainder(): string {
+    return this.index === 0
+      ? this.#input.value
+      : this.#input.value.slice(this.index);
+  }
+
+  /** Mutate the index and return this Cursor. */
+  move(amount: number): this {
+    this.index += amount;
+    return this;
+  }
+
+  /** Shorthand for `this.input.startsWith(str, this.index)`. */
+  startsWith(str: string): boolean {
+    return this.#input.value.startsWith(str, this.index);
+  }
+
+  toString(): string {
+    return `Cursor(${JSON.stringify(this.input)}, ${this.index})`;
+  }
+
+  /** The function called by `console.log()` in Deno. */
+  [Symbol.for("Deno.customInspect")](): string {
+    return this.toString();
```

Comment:
I want to update this to be magical - when logging a `Cursor`...

- A portion of the *current line* should be logged

  - This would benefit from `location()` (#78) as well as caching `lines` for use within `location()`. The string that `Cursor` wraps will never change, thus its lines will never change, and for long inputs

- That portion should span up to the terminal width, using ellipses to convey truncation, similar to the effect in `evalCodeBlocks`

  - Seems we need an `elide()` helper (elide = create elision, as in ellipses) which we can also use in `evalCodeBlocks`

  - May want a `span()` helper? We'll need to create a substring centered on `Location["column"]` but which does not wrap to the previous line or into negative indices. It may be doable with string methods - if not, we'll want a helper.

- A caret (`^`) should be positioned on a second line below, colored if color is available, pointing to the character at `Location["offset"]`

Example:

```ts
...quick brown fox jumps over the laz...
                 ^
```

Reply:
Most of this was implemented! The rest was deemed out-of-scope. 🧙